python/ql/src/experimental/semmle/python/frameworks/Django.qll

```diff
@@ -627,6 +638,905 @@ private module Django {
+      module response {
+        /**
+         * Gets a reference to the attribute `attr_name` of the `django.http.response` module.
+         * WARNING: Only holds for a few predefined attributes.
+         */
+        private DataFlow::Node response_attr(DataFlow::TypeTracker t, string attr_name) {
+          attr_name in ["HttpResponse",
+                // HttpResponse subclasses
+                "HttpResponseRedirect", "HttpResponsePermanentRedirect", "HttpResponseNotModified",
+                "HttpResponseBadRequest", "HttpResponseNotFound", "HttpResponseForbidden",
+                "HttpResponseNotAllowed", "HttpResponseGone", "HttpResponseServerError",
+                "JsonResponse",
+                // HttpResponse-like classes
+                "StreamingHttpResponse", "FileResponse"] and
+          (
+            t.start() and
+            result = DataFlow::importNode("django.http.response" + "." + attr_name)
+            or
+            t.startInAttr(attr_name) and
+            result = response()
+          )
+          or
+          // Due to bad performance when using normal setup with `response_attr(t2, attr_name).track(t2, t)`
+          // we have inlined that code and forced a join
+          exists(DataFlow::TypeTracker t2 |
+            exists(DataFlow::StepSummary summary |
+              response_attr_first_join(t2, attr_name, result, summary) and
+              t = t2.append(summary)
+            )
+          )
+        }
+
+        pragma[nomagic]
+        private predicate response_attr_first_join(
+          DataFlow::TypeTracker t2, string attr_name, DataFlow::Node res,
+          DataFlow::StepSummary summary
+        ) {
+          DataFlow::StepSummary::step(response_attr(t2, attr_name), res, summary)
+        }
+
+        /**
+         * Gets a reference to the attribute `attr_name` of the `django.http.response` module.
+         * WARNING: Only holds for a few predefined attributes.
+         */
+        private DataFlow::Node response_attr(string attr_name) {
+          result = response_attr(DataFlow::TypeTracker::end(), attr_name)
+        }
+
+        /**
+         * Provides models for the `django.http.response.HttpResponse` class
+         *
+         * See https://docs.djangoproject.com/en/3.1/ref/request-response/#django.http.HttpResponse.
+         */
+        module HttpResponse {
+          /** Gets a reference to the `django.http.response.HttpResponse` class. */
+          private DataFlow::Node classRef(DataFlow::TypeTracker t) {
+            t.start() and
+            result = response_attr("HttpResponse")
+            or
+            // TODO: remove/expand this part of the template as needed
+            // Handle `http.HttpResponse` alias
+            t.start() and
+            result = http_attr("HttpResponse")
+            or
+            // subclass
+            result.asExpr().(ClassExpr).getABase() = classRef(t.continue()).asExpr()
```

Comment:
I feel like we should just include this in all of the models of  classes that extend `HttpResponse`. Would this not do what we want?

Reply:
Yes, I think you are right, and it is not a lot of code to duplicate...